nano/node/active_transactions.cpp

```diff
@@ -742,26 +743,58 @@ bool nano::active_transactions::active (nano::block const & block_a)
 	return active (block_a.qualified_root ());
 }
 
-void nano::active_transactions::update_difficulty (nano::block const & block_a)
+void nano::active_transactions::update_difficulty (std::shared_ptr<nano::block> block_a, boost::optional<nano::write_transaction const &> opt_transaction_a)
 {
-	nano::lock_guard<std::mutex> lock (mutex);
-	auto existing (roots.find (block_a.qualified_root ()));
-	if (existing != roots.end ())
+	nano::unique_lock<std::mutex> lock (mutex);
+	auto existing_election (roots.find (block_a->qualified_root ()));
+	if (existing_election != roots.end ())
 	{
 		uint64_t difficulty;
-		auto error (nano::work_validate (block_a, &difficulty));
+		auto error (nano::work_validate (*block_a, &difficulty));
 		(void)error;
 		assert (!error);
-		if (difficulty > existing->difficulty)
+		if (difficulty > existing_election->difficulty)
 		{
 			if (node.config.logging.active_update_logging ())
 			{
-				node.logger.try_log (boost::str (boost::format ("Block %1% was updated from difficulty %2% to %3%") % block_a.hash ().to_string () % nano::to_string_hex (existing->difficulty) % nano::to_string_hex (difficulty)));
+				node.logger.try_log (boost::str (boost::format ("Block %1% was updated from difficulty %2% to %3%") % block_a->hash ().to_string () % nano::to_string_hex (existing_election->difficulty) % nano::to_string_hex (difficulty)));
 			}
-			roots.modify (existing, [difficulty](nano::conflict_info & info_a) {
+			roots.modify (existing_election, [difficulty](nano::conflict_info & info_a) {
 				info_a.difficulty = difficulty;
 			});
-			adjust_difficulty (block_a.hash ());
+			adjust_difficulty (block_a->hash ());
+		}
+	}
+	else if (opt_transaction_a.is_initialized ())
+	{
+		// Only guaranteed to immediately restart the election if the new block is received within 60s of dropping it
+		static constexpr std::chrono::seconds recently_dropped_cutoff{ 60s };
+		if (find_dropped_elections_cache (block_a->qualified_root ()) > std::chrono::steady_clock::now () - recently_dropped_cutoff)
+		{
+			lock.unlock ();
+			nano::block_sideband existing_sideband;
+			auto hash (block_a->hash ());
+			auto existing_block (node.store.block_get (*opt_transaction_a, hash, &existing_sideband));
+			release_assert (existing_block != nullptr);
+			uint64_t confirmation_height;
+			release_assert (!node.store.confirmation_height_get (*opt_transaction_a, node.store.block_account (*opt_transaction_a, hash), confirmation_height));
+			bool confirmed = (confirmation_height >= existing_sideband.height);
+			if (!confirmed && existing_block->block_work () != block_a->block_work ())
+			{
+				uint64_t existing_difficulty, new_difficulty;
```

Comment:
I've not seen many (if any) areas where we declare multiple variables on the same line, so to be consistent they should probably be on separate lines.